scripts/generate_example_ttl_files.py

```diff
@@ -0,0 +1,120 @@
+    # are consistent between runs of this script,
+    # and are consistent for a particular code path.
+
+    # Check rdflib.BNode.__new__ has signature we expect
+    b_node_signature = inspect.signature(rdflib.BNode.__new__)
+    expected_params = ["cls", "value", "_sn_gen", "_prefix"]
+    actual_params = [p.name for p in b_node_signature.parameters.values()]
+    defaults = rdflib.BNode.__new__.__defaults__
+    if actual_params != expected_params or defaults != (None, None, "N"):
+        raise RuntimeError("BNode.__new__ had unexpected signature, the following code might need to be updated")
+    # Override default _sn_gen function with our custom one.
+    rdflib.BNode.__new__.__defaults__ = (defaults[0], _custom_sn_gen, defaults[2])
+
+    # 2. second override:
+    # Mock "Date.today()" function to always give the same result
+    unittest.mock.patch.object(
+        abis_mapping.types.temporal.Date,
+        "today",
+        new=lambda: abis_mapping.types.temporal.Date(2020, 1, 1),
+    ).start()
+
+    # Finished overrides.
+
+    # Generate new ttl files.
+    for template_id, input_csv_file_path, output_ttl_file_path in examples:
+        print(f"Generating {output_ttl_file_path}...")
+
+        # set global var
+        TTL_FILE_PATH = str(output_ttl_file_path)
+
+        # Get Mapper
+        mapper = abis_mapping.get_mapper(template_id)
+        if mapper is None:
+            raise RuntimeError(f"Mapper not found for {template_id}")
+        # Map data
+        data = input_csv_file_path.read_bytes()
+        graphs = list(mapper().apply_mapping(data))
+        if len(graphs) != 1:
+            raise RuntimeError("apply_mapping did not produce exactly 1 graph")
+        # Write to output file
+        graphs[0].serialize(destination=output_ttl_file_path, format="turtle")
+
+    print("Done!")
+
+
+def _custom_sn_gen() -> str:
+    """Generate a BNode ID depending on the ttl file and the current code path.
+
+    Returns:
+        String to use for a BNode ID. Should be 32 hex digits to match rdflib default behavior.
+    """
+    # Generate ID with 32 hex digits (16 bytes) to match the normal behavior (which is uuid4().hex)
+    # First 12 bytes are from hash of ttl file path and the code path taken to create the BNode().
+    # Remaining 4 bytes are taken from a counter unique to the hash result.
+    # i.e. when the same hash occurs, they will be numbered 0,1,2,etc. at the end of the ID.
+    # This ensures each ID should be globally unique.
+    code_path_hash = hashlib.blake2b(digest_size=12, person=b"gen_b_node_id")
+    # hash ttl file path.
+    code_path_hash.update(TTL_FILE_PATH.encode("utf-8"))
+    # Hash functions in the abis_mapping module called to make this BNode()
+    for frame in inspect.stack():
+        # NOTE "abis_mapping" with underscore is the module name.
+        # NOT "abis-mapping" with a dash which is the repository name.
+        if "abis_mapping" in frame.filename:
```

Comment:
Looking at the output turtle, i don't think this is retrieving the correct frame. I think you would rather inspect the frame.function and look out for `apply_mapping` (or both?)? Again though i don't think you need this at all. A single counter that gets reset each mapping file should be enough, also making this function part of a closure would be a better practice rather than relying on globals for it i think.

Reply:
Multiple frames' `function` are being included in each hash -
from my debugging the effective list of `frame.function`s included in each hash are things like
```python
['original_datum_uri', 'to_rdf_literal', 'add_geometry_supplied_as', 'add_sampling_field', 'apply_mapping_row', 'apply_mapping']
['add_observation_scientific_name', 'apply_mapping_row', 'apply_mapping']
['add_observation_verbatim_id', 'apply_mapping_row', 'apply_mapping']
```
So BNodes created from different parts of the mapping code have different hashes

Reply:
This `if` check is really just to exclude the inner part of the stack - which is rdflib internals, 
and to exclude the outer part of the stack (everything before `apply_mapping`),
which could be different depending on if you're running the script directly, or from a debugger or something